Queue poller has failed! It will stop. ... and never restart

Hi, 

We are running JQM v2.2.8 with Oracle Database 19c Enterprise Edition Release 19.4.0.0.0.
We temporary faced an overload of the number of process on the DB side.  Which cause some of the poller to be stopped with the following error.

Of course, the root cause of this outage is not due to JQM, but would it be possible to allow JQM to recover by itself to such temporary issue ? To have a retry + back-off policy ?

Currently, we need to reboot the node to recover the stopped poller.

Thanks

```
23/08 05:19:15.984|ERROR|R;polling;USR_VIP;GepJqmEngineREC2_01_01|QueuePoller      |Queue poller has failed! It will stop.
com.enioka.jqm.jdbc.DatabaseException: java.sql.SQLException: Listener refused the connection with the following error:
ORA-12516, TNS:listener could not find available handler with matching protocol stack
 
	at com.enioka.jqm.jdbc.Db.getConn(Db.java:525)
	at com.enioka.jqm.tools.Helpers.getNewDbSession(Helpers.java:94)
	at com.enioka.jqm.tools.QueuePoller.run(QueuePoller.java:226)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.sql.SQLException: Listener refused the connection with the following error:
ORA-12516, TNS:listener could not find available handler with matching protocol stack
 
	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:854)
	at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:793)
	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:57)
	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:747)
	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:562)
	at org.apache.tomcat.jdbc.pool.PooledConnection.connectUsingDriver(PooledConnection.java:278)
	at org.apache.tomcat.jdbc.pool.PooledConnection.connect(PooledConnection.java:182)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.createConnection(ConnectionPool.java:730)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.borrowConnection(ConnectionPool.java:664)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.getConnection(ConnectionPool.java:200)
	at org.apache.tomcat.jdbc.pool.DataSourceProxy.getConnection(DataSourceProxy.java:128)
	at com.enioka.jqm.jdbc.Db.getConn(Db.java:508)
	... 3 more
Caused by: oracle.net.ns.NetException: Listener refused the connection with the following error:
ORA-12516, TNS:listener could not find available handler with matching protocol stack
 
	at oracle.net.ns.NSProtocolNIO.negotiateConnection(NSProtocolNIO.java:284)
	at oracle.net.ns.NSProtocol.connect(NSProtocol.java:340)
	at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1596)
	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:588)
	... 14 more
```

--- a/jqm-all/jqm-engine/src/main/java/com/enioka/jqm/engine/Helpers.java
+++ b/jqm-all/jqm-engine/src/main/java/com/enioka/jqm/engine/Helpers.java
@@ -303,12 +303,17 @@ final class Helpers
     static boolean testDbFailure(Exception e)
     {
         Throwable cause = e.getCause();
+        Throwable rootCause = ExceptionUtils.getRootCause(e);
+
         return (ExceptionUtils.indexOfType(e, SQLTransientException.class) != -1)
                 || (ExceptionUtils.indexOfType(e, SQLNonTransientConnectionException.class) != -1)
                 || (ExceptionUtils.indexOfType(e, SocketException.class) != -1)
                 || (ExceptionUtils.indexOfType(e, SocketTimeoutException.class) != -1)
-                || (cause != null && cause.getMessage().equals("This connection has been closed"))
-                || (cause instanceof SQLException && e.getMessage().equals("Failed to validate a newly established connection."))
-                || (cause instanceof SQLNonTransientException && cause.getMessage().equals("connection exception: closed"));
+                || (rootCause != null && rootCause.getClass().getName().equals("oracle.net.ns.NetException"))
+                || (cause != null && cause.getMessage() != null && cause.getMessage().equals("This connection has been closed"))
+                || (cause instanceof SQLException && e.getMessage() != null
+                        && e.getMessage().equals("Failed to validate a newly established connection."))
+                || (cause instanceof SQLNonTransientException && cause.getMessage() != null
+                        && cause.getMessage().equals("connection exception: closed"));
     }
 }

